VetLife/Assets/Scripts/UserInput/GestureHandler.cs

```diff
@@ -0,0 +1,762 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace Assets.Scripts.UserInput
+{
+	#region Gesture Gatherer State Machine
+
+	/// <summary>
+	/// Represents base abstract gesture gatherer state
+	/// </summary>
+	internal abstract class GathererState
+	{
+		#region Properties
+
+		/// <summary>
+		/// Reference to the <see cref="GestureHandler"/> owner object
+		/// </summary>
+		protected GestureHandler Handler;
+
+		/// <summary>
+		/// The time current state started
+		/// </summary>
+		protected readonly float StartTime;
+
+		#endregion
+
+		#region Constructors
+
+		/// <summary>
+		/// Constructs base gathering state
+		/// </summary>
+		/// <param name="handler">Reference to the <see cref="GestureHandler"/> owner object</param>
+		protected GathererState( GestureHandler handler )
+		{
+			Handler = handler;
+			StartTime = Time.time;
+		}
+
+		#endregion
+
+		#region Functions
+
+		/// <summary>
+		/// Responds to situation when there is new unregistered input
+		/// </summary>
+		internal virtual void OnNewInput()
+		{
+			if( Handler.ActiveGesture != null )
+			{
+				Handler.FinishGesture();
+			}
+
+			Handler.ChangeState( new GatheringState( Handler ) );
+		}
+
+		/// <summary>
+		/// Handles the updating of currently active touches and gesture
+		/// </summary>
+		internal abstract void OnUpdate();
+
+		/// <summary>
+		/// Checks, whether time elapsed in this state has crossed given threshold
+		/// </summary>
+		/// <param name="threshold">Threshold to check</param>
+		/// <returns>True in case state is up longer than given threshold, false otherwise</returns>
+		internal bool HasTimePassed( float threshold )
+		{
+			return Time.time - StartTime >= threshold;
+		}
+
+		/// <summary>
+		/// Checks, whether given <see cref="Touch"/> is stationary compared to last frame
+		/// </summary>
+		/// <param name="touch"><see cref="Touch"/> to be checked</param>
+		/// <returns>True in case the delta of <see cref="Touch"/> position is lesser than given threshold, false otherwise</returns>
+		internal bool IsStationary( Touch touch )
+		{
+			return touch.deltaPosition.magnitude <= Handler.StationaryDistanceThreshold;
+		}
+
+		#endregion
+	}
+
+	/// <summary>
+	/// Represents gatherer state, when no input is being processed
+	/// </summary>
+	internal class IdleState : GathererState
+	{
+		#region Constructors
+
+		/// <summary>
+		/// Constructs base idle state
+		/// </summary>
+		/// <param name="handler">Reference to the <see cref="GestureHandler"/> owner object</param>
+		internal IdleState( GestureHandler handler ) : base( handler )
+		{
+			// intentionally left blank
+		}
+
+		#endregion
+
+		#region Overrides
+
+		internal override void OnUpdate()
+		{
+			// intentionally left blank
+		}
+
+		#endregion
+	}
+
+	/// <summary>
+	/// Represents gatherer state, which input is being gathered in
+	/// </summary>
+	internal class GatheringState : GathererState
```

Comment:
A bit of a nitpick: Wouldn't it be better to name either this or its parent class a little more distinctly?

Reply:
Well, you suggest it would be better, but this is the best I was able to come up with, so I doubt I will come up with better ones now ... 

The whole process gathers input - therefore _GathererState_. However, only one state in the entire SM gathers input; those others collect additional data for gestures - therefore _GatheringState_.

Also part of constructive criticism is to come up with ideas, so if you would provide some, I would be glad to think about them O:))

Reply:
Maybe the parent object could be called something like `InputState`, since it deals with inputs in general, not just the gathering of them.  Either way, if it stays the way it is I don't think that the name is used in many contexts where it's not pretty easy to figure out which is which.